clang/lib/Sema/SemaTemplateDeduction.cpp

```diff
@@ -3553,6 +3553,49 @@ static unsigned getPackIndexForParam(Sema &S,
   llvm_unreachable("parameter index would not be produced from template");
 }
 
+// if `Specialization` is a `CXXConstructorDecl` or `CXXConversionDecl`
+// we try to instantiate and update its explicit specifier after constraint
+// checking.
+static Sema::TemplateDeductionResult
+resolveExplicitSpecifier(Sema &S, FunctionDecl *Specialization,
+                         const MultiLevelTemplateArgumentList &SubstArgs,
+                         TemplateDeductionInfo &Info,
+                         FunctionTemplateDecl *FunctionTemplate,
+                         ArrayRef<TemplateArgument> DeducedArgs) {
+  auto GetExplicitSpecifier = [](FunctionDecl *D) {
+    return isa<CXXConstructorDecl>(D)
+               ? cast<CXXConstructorDecl>(D)->getExplicitSpecifier()
+               : cast<CXXConversionDecl>(D)->getExplicitSpecifier();
+  };
+  auto SetExplicitSpecifier = [](FunctionDecl *D, ExplicitSpecifier ES) {
+    isa<CXXConstructorDecl>(D)
+        ? cast<CXXConstructorDecl>(D)->setExplicitSpecifier(ES)
+        : cast<CXXConversionDecl>(D)->setExplicitSpecifier(ES);
+  };
+
+  ExplicitSpecifier ES = GetExplicitSpecifier(Specialization);
+  Expr *const Expr = ES.getExpr();
+  if (!Expr) {
+    return Sema::TDK_Success;
+  }
+  if (!Expr->isValueDependent()) {
+    return Sema::TDK_Success;
+  }
+  // TemplateDeclInstantiator::InitFunctionInstantiation set the
+  // ActiveInstType to TemplateInstantiation, but we need
+  // to enable SFINAE when instantiating an explicit specifier.
+  Sema::InstantiatingTemplate Inst(
```

Comment:
This part is a bit tricky... Actually, the `InstantiatingTemplate` here is only used to enable SFINAE, the instantiation depth/count is the same as `FinishTemplateArgumentDeduction`. If the `InstantiatingTemplate` in `FinishTemplateArgumentDeduction` is valid, the `InstantiatingTemplate` in `resolveExplicitSpecifier` should be valid, too.

Is that theory correct? If that's correct, I'll add comment for this.

Reply:
> the InstantiatingTemplate here is only used to enable SFINAE...

I'm wondering if [`SFINAETrap`](https://github.com/llvm/llvm-project/blob/e6971e5a41fe30264af9a13c8f387c06f93c6d9c/clang/include/clang/Sema/Sema.h#L10006-L10007) does the trick for you.

Reply:
Thanks for reminding :)
I think you are right,`SFINAETrap` seems do the trick here. I'll take a deeper look at this.

Reply:
I'm a bit confused here. Do I need to check `SFINAETrap::hasErrorOccurred`? Is `ExprResult::IsValid` enough?

Reply:
I would suggest checking both?  But I cannot come up with a reason why just the ExprResult's validity wouldn't be sufficient.

Reply:
And yet, as soon as I said that, I remembered:  a `RecoveryExpr` (or other attempts at error recovery) can result in an error happening, but a valid expression being returned.

Reply:
Thanks for your reminder! I've added the `hasErrorOccurred` check.
Also I renamed the function because I found a function named `tryResolveExplicitSpecifier` and I want to make them distinct.

Reply:
> So we don't have to check Inst.isInvalid()?

comments added

Reply:
Sorry, why do not have to check `Inst.isInvalid`?  We do it everywhere, and it is a good 'catch' for mistakes.

Reply:
The instantiation depth in `instantiateExplicitSpecifierDeferred` is the depth in `FinishTemplateArgumentDeduction` plus one, but I think, actually, their depths should be equal ... If the depth in `FinishTemplateArgumentDeduction` is valid, the depth in `instantiateExplicitSpecifierDeferred` should be valid, too. So, I didn't check `IsValid`: If I check it, the available depth will be subtracted by one

But it seems that `IsValid` not only check the instantiation depth ... :(

Reply:
Right, there is quite a bit that it checks, so we  ALWAYS check it (or at least should).

Reply:
ok :)